packages/ddp-server/livedata_server.js

```diff
@@ -1643,25 +1653,34 @@ _.extend(Server.prototype, {
       throw new Error("Can't call setUserId on a server initiated method call");
     };
     var connection = null;
-    var currentInvocation = DDP._CurrentInvocation.get();
-    if (currentInvocation) {
-      userId = currentInvocation.userId;
+    var currentMethodInvocation = DDP._CurrentMethodInvocation.get();
+    var currentPublicationInvocation = DDP._CurrentPublicationInvocation.get();
+    var randomSeed = null;
+    if (currentMethodInvocation) {
+      userId = currentMethodInvocation.userId;
+      setUserId = function(userId) {
+        currentMethodInvocation.setUserId(userId);
+      };
+      connection = currentMethodInvocation.connection;
+      randomSeed = DDPCommon.makeRpcSeed(currentMethodInvocation, name);
+    } else if (currentPublicationInvocation) {
+      userId = currentPublicationInvocation.userId;
       setUserId = function(userId) {
-        currentInvocation.setUserId(userId);
+        currentPublicationInvocation._session._setUserId(userId);
```

Comment:
I also looked into this because it was suspicious to me, but I could not find a cleaner way either. It seems the best and most stable one.

Maybe we should make a test which makes sure you call `setUserId` from inside a method called from publish function to make sure this will be always available. It would also in general be interesting to see what happens if you call `setUserId` from inside a publish function.